File: Java/other/RainWaterHarvesting.java
import java.util.Scanner;

class RainWaterHarvesting {

    public static int saveWater(int[] arr, int size) {

        int result = 0;

        // traverse the array from second to second last element
        for (int i = 1; i < size - 1; i++) {

            // Find maximum element on its left
            int left = arr[i];
            for (int j = 0; j < i; j++) {
                left = Math.max(left, arr[j]);
            }

            // Find maximum element on its right
            int right = arr[i];
            for (int j = i + 1; j < size; j++) {
                right = Math.max(right, arr[j]);
            }

            // Update maximum water value
            result += Math.min(left, right) - arr[i];
        }
        return result;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.println("Enter array size:");
        int size = scanner.nextInt();
        int[] arr = new int[size];
        System.out.println("Enter array elements:");
        for (int i = 0; i < size; i++) {
            arr[i] = scanner.nextInt();
        }
        scanner.close();

        System.out.println("Maximum amount of water that can be saved is:" + saveWater(arr, size));
    }
}

/*
* Sample input/output:
* Enter array size:
* 10
* Enter array elements:
* 0 2 1 3 0 1 2 1 2 1
* Maximum amount of water that can be saved is:5
*
* Time complexity: O(n^2)
* space complexity: O(1)
*/

File: C-Plus-Plus/other/Rainwater_harvesting.cpp
/*
Given the size of the array and its element. We have to compute the 
amount of water that can be stored in every element of the array.
*/

#include <iostream>
using namespace std;
int maxWater_stored(int a[], int n)
{
    int water_stored = 0; 
    // To store the final ans
    int left_max = 0; 
    // Which stores the current max height of the left side
    int right_max = 0; 
    // Which stores the current max height of the right side
    int low = 0; 
    // Counter to traverse from the left_side
    int high = n - 1;
     // Counter to traverse from the right_side
    while (low <= high)
    {
        if (a[low] < a[high])
        {
            if (a[low] > left_max)
                left_max = a[low]; 
                // Updating left_max
            else
                water_stored += left_max - a[low]; 
                // Calculating the ans
            low++;
        }
        else
        {
            if (a[high] > right_max)
                right_max = a[high]; 
                // Updating right_max
            else
                water_stored += right_max - a[high]; 
                // Calculating the ans
            high--;
        }
    }
    return water_stored;
}
int main() {
    int n;
    //storing the size of the array
    cout<<"Enter the size of array : ";
    cin>>n;
    int a[n];
    //storing the array
    cout<<"Enter the array elements : ";
    for(int i=0;i<n;i++)
    cin>>a[i];
    //calling the function and storing the answer
    int ans=maxWater_stored(a,n);
    cout<<"The amount of water stored is "<<ans<<" units";
    return 0;
}

/*
Sample Input :
Enter the size of array : 6
Enter the array elements : 3 0 0 2 0 4
Output :
The amount of water stored is 10 units
*/

/*
Time Complexity : O(n)
Space Complexity : O(1)
*/
